### PR TITLE
ISPN-5962 Implementation of RpcManagerImpl.invokeRemotely causes high CPU usage

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.jboss.org/browse/ISPN-5962

did some cleanups as well.

### CODE REVIEW

File: core/src/main/java/org/infinispan/remoting/rpc/RpcManager.java [hunk 1]
```diff
@@ -148,4 +150,17 @@ void invokeRemotelyInFuture(NotifyingNotifiableFuture<Map<Address, Response>> fu
     */
    RpcOptions getDefaultRpcOptions(boolean sync, DeliverOrder deliverOrder);
 
+   static Map<Address, Response> waitFor(CompletableFuture<Map<Address, Response>> future, RpcOptions options)
+         throws InterruptedException, ExecutionException {
+      long waitTime = options.timeout() * 2;
```
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/remoting/rpc/RpcManager.java]: you have this exact same logic in https://github.com/infinispan/infinispan/pull/3841/files#diff-36bedadd971913e8bd8c08cc2e79e9a2R934 below. Maybe coalesce them ?

**Comment by pruivo** [comment]
@tristantarrant updated. I've created a ```RpcUtil``` and I put the method there. It is used everywhere (```RpcManager``` and ```Transport```). let me know that you think.

File: core/src/main/java/org/infinispan/remoting/rpc/RpcManagerImpl.java [hunk 2]
```diff
@@ -134,14 +134,7 @@ private boolean useReplicationQueue(boolean sync) {
...
       // Set the topology id of the command, in case we don't have it yet
-      if (rpc instanceof TopologyAffectedCommand) {
-         TopologyAffectedCommand topologyAffectedCommand = (TopologyAffectedCommand) rpc;
-         if (topologyAffectedCommand.getTopologyId() == -1) {
-            int currentTopologyId = stateTransferManager.getCacheTopology().getTopologyId();
-            if (trace) log.tracef("Topology id missing on command %s, setting it to %d", rpc, currentTopologyId);
-            topologyAffectedCommand.setTopologyId(currentTopologyId);
-         }
-      }
+      setTopologyId(rpc);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/remoting/rpc/RpcManagerImpl.java]: I'd rather wait for a constant time, expressed in nanos so there's no conversion during the RPC (e.g. `private static final long BIG_DELAY_NANOS = TimeUnit.DAYS.toNanos(1)`).

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManagerImpl.java [hunk 3]
```diff
@@ -365,14 +367,14 @@ private BackupRpcConfiguration backupRpcConfiguration(String siteName) {
...
-            command, rpcManager.getDefaultRpcOptions(true, DeliverOrder.NONE));
+      RpcOptions options = rpcManager.getDefaultRpcOptions(true, DeliverOrder.NONE);
+      CompletableFuture<Map<Address, Response>> remoteFuture = rpcManager.invokeRemotelyAsync(null, command, options);
       final Future<Response> localFuture = asyncExecutor.submit(
             LocalInvocation.newInstance(responseGenerator, command, commandsFactory, rpcManager.getAddress()));
       final Map<Address, Response> responseMap = new HashMap<>();
       responseMap.put(rpcManager.getAddress(), localFuture.get());
       //noinspection unchecked
-      responseMap.putAll(remoteFuture.get());
+      responseMap.putAll(RpcUtil.waitFor(remoteFuture, options.timeout(), options.timeUnit()));
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManagerImpl.java]: Could you make this async and send the reply from `remoteFuture.whenCompleted()` instead?

Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManagerImpl.java]: I don't see the benefit of it.

File: core/src/main/java/org/infinispan/util/RpcUtil.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * @since 8.1
+ */
+public class RpcUtil {
+
+   public static final long BIG_DELAY_NANO = TimeUnit.DAYS.toNanos(1);
+
+   private RpcUtil() {
+   }
+
+   public static <T> T waitFor(CompletableFuture<T> future)
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/RpcUtil.java]: One more (very late) suggestion: this should be useful for anything that needs to wait on a `CompletableFuture`, so I'd move it to `CompletableFutures`.

**Comment by pruivo** [comment]
@danberindei updated!

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java [hunk 5]
```diff
@@ -502,7 +502,7 @@ public Address getAddress() {
       CompletableFuture<Map<Address, Response>> future = invokeRemotelyAsync(recipients, rpcCommand, mode,
             timeout, responseFilter, deliverOrder, anycast);
       try {
-         return future.get();
+         return CompletableFutures.await(future);
```
Comment (galderz) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: Could we add some comments here why we can afford to wait endlessly here? What is providing the guarantees that we are sure that we'll never wait endlessly?

Comment (pruivo) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: what do you mean by "wait endlessly"? all the rpc has a timeout.

Comment (danberindei) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: @pruivo Please add a comment saying the `invokeRemotelyAsync` is guaranteed to complete within `timeout` millis, so there's no need to use a timeout here.

**Comment by pruivo** [comment]
add the comment and rebased @galderz and @danberindei .

**Comment by danberindei** [comment]
Integrated, thanks Pedro!

**Pull request closed without merging**